### PR TITLE
Move Tuple and Either types to DA.Types

### PR DESCRIPTION
**Pull request by neil-da**
Move Tuple* and Either to daml-prim:DA.Types, so they have a nice name for Java users (instead of GHC.Tuple or DA.Internal.Prelude). Note that this will break anyone using the Java ecosystem since their location will have changed.

Requested by @jberthold-da. CC @bitonic

### CODE REVIEW

File: daml-foundations/daml-ghc/daml-prim-src/DA/Types.daml [hunk 1]
```diff
@@ -0,0 +1,60 @@
+{-# LANGUAGE NoImplicitPrelude #-}
+
+daml 1.2
+
+-- | A module containing all the standard types from the base libraries,
+--   so they have nice names when used from Java or similar.
+module DA.Types where
```
Comment (neil-da) [inline daml-foundations/daml-ghc/daml-prim-src/DA/Types.daml]: My biggest question was should this be DA.Type or DA.Types... I went with the latter but open to opposing opinions.

Comment (hurryabit) [reply daml-foundations/daml-ghc/daml-prim-src/DA/Types.daml]: I think `DA.Types` is good. `DA.Type` would sound like there's one type called `Type` in there.

File: daml-foundations/daml-ghc/daml-prim-src/DA/Types.daml [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+-- type `Either a b` is either `Left a` or `Right b`.
+--
+-- The `Either` type is sometimes used to represent a value which is
+-- either correct or an error; by convention, the `Left` constructor is
+-- used to hold an error value and the `Right` constructor is used to
+-- hold a correct value (mnemonic: "right" also means "correct").
+data Either a b = Left a | Right b
+
+
+-- NOTE(MH): These types are not supposed to be by DAML users. The converter to
```
Comment (hurryabit) [inline daml-foundations/daml-ghc/daml-prim-src/DA/Types.daml]:
```suggestion
-- NOTE(MH): These types are not supposed to be used in DAML. The converter to
```

File: daml-foundations/daml-ghc/daml-prim-src/DA/Types.daml (hunk 1 above)
Comment (hurryabit) [inline daml-foundations/daml-ghc/daml-prim-src/DA/Types.daml]:
```suggestion
module DA.Types
  ( Either (..)
  ) where
```
The `TupleN` types are not supposed to be used in DAML anyway and DAML-LF does not care about export statements.

File: docs/source/support/release-notes.rst [hunk 3]
```diff
@@ -9,6 +9,9 @@ This page contains release notes for the SDK.
 HEAD — ongoing
 --------------
 
+- **DAML Standard Library**: Move the ``Tuple`` and ``Either`` types to ``daml-prim:DA.Types`` rather
+  than exposing internal locations
```
Comment (hurryabit) [inline docs/source/support/release-notes.rst]: We should also leave an explanation how users need to update their applications on the Java/Scala/... side. cc @bethaitman

Comment (hurryabit) [reply docs/source/support/release-notes.rst]: We also need to update https://github.com/digital-asset/daml/blob/master/docs/source/app-dev/daml-lf-translation.rst#tuple-types

File: docs/source/support/release-notes.rst [hunk 4]
```diff
@@ -9,6 +9,10 @@ This page contains release notes for the SDK.
 HEAD — ongoing
 --------------
 
+- **DAML Standard Library**: Move the ``Tuple`` and ``Either`` types to ``daml-prim:DA.Types`` rather
+  than exposing internal locations. People using the Scala/Java bindings will need to change from
```
Comment (bethaitman) [inline docs/source/support/release-notes.rst]: Can you rephrase as an instruction? Just to make it a bit more direct, so it's totally obvious what you need to do. Something like "If you're using the Scala or Java bindings, you need to change..." and then what and where - is this in your imports, or what?

Comment (bethaitman) [reply docs/source/support/release-notes.rst]: and this doesn't affect your DAML code/imports at all?

Comment (neil-da) [reply docs/source/support/release-notes.rst]: That would be great, but I have very little idea what you actually need to do.

Comment (neil-da) [reply docs/source/support/release-notes.rst]: No impact on DAML imports at all - they were always being reexposed in other places and they still are

Comment (bethaitman) [reply docs/source/support/release-notes.rst]: If you don't know what users need to do as a result of this change, can you find out, or is that too much of a pain?

Comment (bethaitman) [reply docs/source/support/release-notes.rst]: (and for the DAML imports - it'd be nice to say explicitly "You don't need to change anything in your DAML code)

Comment (neil-da) [reply docs/source/support/release-notes.rst]: More nuanced than that - I did fix up some instances, and they were just replacing `GHC.Tuple` with `DA.Tuple`, so I believe if my instructions are interpreted as a list of replacements that's sufficient, at least for our test suite. What I don't know is if there will be more subtle impacts or other places that need tweaks. I don't think so. I also can't say whether it's in import statements or usages, because that depends on what users have done.

Comment (neil-da) [reply docs/source/support/release-notes.rst]: Cleaned it up a bit

Comment (bethaitman) [reply docs/source/support/release-notes.rst]: I guess this is what I'd suggest, then:

```
- **BREAKING CHANGE - DAML Standard Library**: Moved the ``Tuple`` and ``Either`` types to ``daml-prim:DA.Types`` rather than exposing internal locations.

  You don't need to change DAML code as a result of this change. However if you're using the Java or Scala bindings, you need to migrate by replacing usages or imports of ``GHC.Tuple`` and ``DA.Internal.Prelude`` with ``DA.Tuple``.
```

**Comment by bitonic** [comment]
think hard about the location because we won't be able to move them often!

**Comment by neil-da** [comment]
@bitonic hopefully this is the last ever move. The original location was more happenstance, and didn't say anything sensible and relied on lots of internal details.

File: docs/source/support/release-notes.rst (hunk 4 above)
Comment (neil-da) [reply docs/source/support/release-notes.rst]: I went with that.

Comment (hurryabit) [reply docs/source/support/release-notes.rst]: I had a quick chat with @gerolf-da:
- People using the Java/Scala codegen need to replace `import ghc.tuple.*` with `import da.types.*`. Same for `DA.Internal.Prelude`.
- People using the Ledger API directly need to do the change mentioned above.

Comment (neil-da) [reply docs/source/support/release-notes.rst]: Thanks both, I've integrated that update into the notes.

File: docs/source/support/release-notes.rst [hunk 5]
```diff
@@ -12,6 +12,15 @@ HEAD — ongoing
 DAML
 ~~~~
 
+- **BREAKING CHANGE - DAML Standard Library**: Moved the ``Tuple`` and ``Either`` types to ``daml-prim:DA.Types``
+  rather than exposing internal locations.
+
+How to migrate:
```
Comment (gerolf-da) [inline docs/source/support/release-notes.rst]:
```suggestion
  How to migrate:
```
Adding the leading spaces makes the "migration notes" look more as part of the bullet point it belongs to.

Comment (neil-da) [reply docs/source/support/release-notes.rst]: I was copying from the Ledger API section a few lines below which has made this mistake. Should I fix that too?

Comment (gerolf-da) [reply docs/source/support/release-notes.rst]: Yeah, that was my mistake. Sorry. I'm fixing it in #1107.

**Comment by jberthold-da** [comment]
21 commits, what a ride!
Thanks for actioniing this. I am convinced this makes the APIs much cleaner (worth the pain in the next upgrade).
